final-project/application.py

```diff
@@ -0,0 +1,254 @@
+# Final Project | Application | (jwq1)
+import os
+import re
+from cs50 import SQL
+
+from flask import Flask, jsonify, render_template, request, url_for, flash, session
+from flask_session import Session
+from flask_jsglue import JSGlue
+
+from passlib.apps import custom_app_context as pwd_context
+from tempfile import mkdtemp
+
+from helpers import *
+
+# configure application
+app = Flask(__name__)
+JSGlue(app)
+
+# ensure responses aren't cached
+if app.config["DEBUG"]:
+    @app.after_request
+    def after_request(response):
+        response.headers["Cache-Control"] = "no-cache, no-store, must-revalidate"
+        response.headers["Expires"] = 0
+        response.headers["Pragma"] = "no-cache"
+        return response
+
+# configure session to use filesystem (instead of signed cookies)
+app.config["SESSION_FILE_DIR"] = mkdtemp()
+app.config["SESSION_PERMANENT"] = False
+app.config["SESSION_TYPE"] = "filesystem"
+Session(app)
+
+# configure CS50 Library to use SQLite database
+db = SQL("sqlite:///final.db")
+
+@app.route("/")
+@login_required
+def index():
+
+    # keep user logged in
+    session.get("user_id")
+
+    """Render map."""
+    if not os.environ.get("API_KEY"):
+        raise RuntimeError("API_KEY not set")
+    return render_template("index.html", key=os.environ.get("API_KEY"))
+
+# @app.route("/articles", methods=["GET", "POST"])
+
+@app.route("/login", methods=["GET", "POST"])
+def login():
+    """Log user in."""
+
+    # forget any user_id
+    session.clear()
+
+    # if user reached route via POST (as by submitting a form via POST)
+    if request.method == "POST":
```

Comment:
There is something about single-purpose methods, but you can look up that principle.

Here the method would be doing 2 things, so you can extract out those differences into separate methods to reduce method scope.

Reply:
Yeah, I should break all this work out into individual methods more. I will adhere to this more often.